File: chapter_4_trees_graphs/js/tree.js
class TreeNode {
  /* Tree Node constructor */
  constructor(value) {
    this.value = value
    this.left = null
    this.right = null
    this.parent = null
  }

  addLeft(value) {
    /* Adds a node to the left */
    const new_node = new TreeNode(value)
    new_node.parent = this
    this.left = new_node
    return new_node
  }

  addRight(value) {
    /* Adds a node to the right */
    const new_node = new TreeNode(value)
    new_node.parent = this
    this.right = new_node
    return new_node
  }
}

/* Prints current node before child nodes */
const preOrderTraversal = node => {
  if (!node) return
  console.log(node.value)
  preOrderTraversal(node.left)
  preOrderTraversal(node.right)
}

/* Prints left nodes, current node, and then right nodes */
const inOrderTraversal = node => {
  if (!node) return
  inOrderTraversal(node.left)
  console.log(node.value)
  inOrderTraversal(node.right)
}

/* Visits child nodes before current node */
const postOrderTraversal = node => {
  if (!node) return
  postOrderTraversal(node.left)
  postOrderTraversal(node.right)
  console.log(node.value)
}

module.exports = {
  TreeNode,
  preOrderTraversal,
  inOrderTraversal,
  postOrderTraversal
}

File: chapter_4_trees_graphs/python/tree.py
class TreeNode:
    def __init__(self, value):
        """
        Tree constructor
        """
        self.value = value
        self.parent = None
        self.left = None
        self.right = None


def pre_order_traversal(node):
    """
    Prints current node before child nodes
    """
    if not node:
        return
    print(node.value)
    pre_order_traversal(node.left)
    pre_order_traversal(node.right)


def in_order_traversal(node):
    """
    Prints left nodes, current node, and then right nodes
    """
    if not node:
        return
    in_order_traversal(node.left)
    print(node.value)
    in_order_traversal(node.right)


def post_order_traversal(node):
    """
    Visits child nodes before current node
    """
    if not node:
        return
    post_order_traversal(node.left)
    post_order_traversal(node.right)
    print(node.value)
